src/AccountManager.ts

```diff
@@ -0,0 +1,134 @@
+  | EnvelopingTypes.DeployRequestStruct;
+
+export default class AccountManager {
+  private _provider: providers.Provider;
+
+  private _signer: Signer;
+
+  private _accounts: Wallet[] = [];
+
+  chainId: number;
+
+  constructor(signer: Signer, provider: providers.Provider, chainId: number) {
+    this._signer = signer;
+    this._provider = provider;
+    this.chainId = chainId;
+  }
+
+  getAccounts(): string[] {
+    return this._accounts.map((it) => it.address);
+  }
+
+  addAccount(account: Wallet): void {
+    const wallet = new Wallet(account.privateKey, this._provider);
+    if (wallet.address !== account.address) {
+      throw new Error('invalid keypair');
+    }
+    this._accounts.push(wallet);
+  }
+
+  isDeployRequest({ request }: RelayOrDeployRequest): boolean {
+    return 'recoverer' in request;
+  }
+
+  async sign(relayRequest: RelayOrDeployRequest): Promise<string> {
+    const cloneRequest = { ...relayRequest }; // why does it have to be clone? don't seem to be mutated.
+    const {
+      relayData: { callForwarder },
+      request: { from },
+    } = relayRequest as typeof relayRequest & {
+      relayData: { callForwarder: string };
+      request: { from: string };
+    }; // this is here just to remove PromiseOrValue<string> type from the def string properties
+
+    const isDeploy = this.isDeployRequest(relayRequest);
+
+    const data = getEnvelopingRequestDataV4Field({
+      chainId: this.chainId,
+      verifier: callForwarder,
+      relayRequest: cloneRequest,
+      requestTypes: isDeploy ? deployRequestType : relayRequestType,
+    });
+
+    const wallet = this._accounts.find(
+      (account) => getAddress(account.address) === getAddress(from)
+    );
+    let rec: string;
+    let signature: string;
+
+    try {
+      if (wallet) {
+        signature = this._signWithWallet(wallet, data);
+      } else {
+        signature = await this._signWithProvider(data);
+      }
+      rec = this._recoverSignature(data, signature);
+    } catch (error) {
+      throw new Error(
+        `Failed to sign relayed transaction for ${from}: ${error as string}`
+      );
+    }
+    if (getAddress(rec) !== getAddress(from)) {
+      throw new Error(
+        `Internal RelayClient exception: signature is not correct: sender=${from}, recovered=${rec}`
+      );
+    }
+
+    return signature;
+  }
+
+  private _recoverSignature(
+    data: TypedMessage<EnvelopingMessageTypes>,
+    signature: string
+  ) {
+    return sigUtil.recoverTypedSignature<
+      SignTypedDataVersion.V4,
+      EnvelopingMessageTypes
+    >({
+      data,
+      signature,
+      version: SignTypedDataVersion.V4,
+    });
+  }
+
+  private async _signWithProvider<T>(
+    data: TypedMessage<EnvelopingMessageTypes>
+  ): Promise<T> {
+    if (!(this._provider as providers.JsonRpcProvider).send) {
+      throw new Error(`Not an RPC provider`);
+    }
+
+    return (await (this._provider as providers.JsonRpcBatchProvider).send(
```

Comment:
```suggestion
    return (await (this._provider as providers.JsonRpcBatchProvider).connect(this._signer).send(
```

Reply:
Not sure about this id which is just a date. Can anyone clarify whether it is used for anything?

![Screenshot 2022-11-01 at 2 05 23 pm](https://user-images.githubusercontent.com/36888576/199252342-8a8c9777-eb75-467a-b23a-0425f7f1b425.png)

Reply:
As far of my understanding. The Json RPC requests always needs an Id, which is meaningless for the blockchain but it could mean something to us. If they sent the date and its not use in other place, we can send anything.

Reply:
just looked into the implementation of the jsonrpcprovider and it autoincrements it's own local id that begins with 42 lol.